### PR TITLE
Rework xml import export, add two new import/export methods

### PR DESCRIPTION
**Pull request by dawagner**
Unify import/export to/from strings and files under a common abstraction: STL streams.

Fix some typos.

### CODE REVIEW

**Comment by dawagner** [comment]
@OznOg @krocard @gdenneul @clero please review; keep in mind that this is an RFC/Work In Progress.

File: parameter/ParameterMgr.cpp [hunk 1]
```diff
@@ -146,157 +148,159 @@ const CParameterMgr::SRemoteCommandParserItem CParameterMgr::gastRemoteCommandPa
             "Show current status" },
 
     /// Tuning Mode
-    { "setTuningMode", &CParameterMgr::setTuningModeCommmandProcess, 1,
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: nice caaaaaaaaaaaaaaaaaatch

File: parameter/ParameterMgr.cpp [hunk 2]
```diff
@@ -519,7 +523,9 @@ bool CParameterMgr::loadFrameworkConfiguration(string& strError)
     // Parse Structure XML file
     CXmlElementSerializingContext elementSerializingContext(strError);
 
-    if (!xmlParse(elementSerializingContext, getFrameworkConfiguration(), _strXmlConfigurationFilePath, _strXmlConfigurationFolderPath, EFrameworkConfigurationLibrary)) {
+    const ifstream file(_strXmlConfigurationFilePath.c_str());
+    // TODO: test file.good()
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: test it then ?

File: parameter/ParameterMgr.cpp [hunk 3]
```diff
@@ -571,7 +577,9 @@ bool CParameterMgr::loadStructure(string& strError)
 
     CAutoLog autolog(pSystemClass, "Importing system structure from file " + strXmlStructureFilePath);
 
-    if (!xmlParse(parameterBuildContext, pSystemClass, strXmlStructureFilePath, strXmlStructureFolder, EParameterCreationLibrary)) {
+    const ifstream file(strXmlStructureFilePath.c_str());
+    // TODO: test file.good()
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: idem

File: parameter/ParameterMgr.cpp [hunk 4]
```diff
@@ -2190,13 +2198,8 @@ bool CParameterMgr::importDomainsXml(const string& strXmlSource, bool bWithSetti
...
     }
 
-    // check path is absolute
-    if (bFromFile && strXmlSource[0] != '/') {
-
-        strError = "Please provide absolute path";
-
-        return false;
-    }
+    CAutoLog autoLog(this, "Importing domains from " +
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: I would put the filename between quotes to clearly see when empty

File: parameter/ParameterMgr.cpp [hunk 5]
```diff
@@ -2301,29 +2282,41 @@ bool CParameterMgr::serializeElement(string& strXmlDest,
                                      strXmlSchemaFilePath, "parameter-framework",
                                      getVersion(), _bValidateSchemasOnStart);
 
-    // Xml Sink
-    CXmlDocSink* pSink;
+    // Use a doc sink to write the doc data in a stream
+    CXmlStreamDocSink sink(xmlDest);
 
-    if (bToFile) {
+    bool bProcessSuccess = sink.process(memorySource, xmlSerializingContext);
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: return sink.process(memorySource, xmlSerializingContext);

File: parameter/ParameterMgr.cpp [hunk 6]
```diff
@@ -2301,29 +2282,41 @@ bool CParameterMgr::serializeElement(string& strXmlDest,
...
+        output.reset(new ofstream(strXmlDest.c_str()));
     } else {
+        output.reset(new ostringstream);
+    }
 
-        // Use a doc sink to write the doc data in a string
-        // TODO: use a stream rather than a string
-        pSink = new CXmlStringDocSink(strXmlDest);
+    // Do the actual export
+    if (!exportDomainsXml(*output, bWithSettings, strError)) {
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: does this leak?

File: parameter/ParameterMgr.cpp [hunk 7]
```diff
@@ -2333,14 +2326,41 @@ bool CParameterMgr::exportDomainsXml(string& strXmlDest, bool bWithSettings, boo
...
+    CAutoLog autoLog(this, "Exporting single domain '" + strDomainName + "' to " +
+            (bToFile ? strXmlDest : "user-provided string"));
+
+    std::auto_ptr<ostream> output(NULL);
+
+    if (bToFile) {
+        output.reset(new ofstream(strXmlDest.c_str()));
+    } else {
+        output.reset(new ostringstream);
+        /* ostringstream* oss(new ostringstream); */
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: dead code

File: parameter/XmlFileIncluderElement.cpp [hunk 8]
```diff
@@ -68,20 +69,21 @@ bool CXmlFileIncluderElement::fromXml(const CXmlElement& xmlElement, CXmlSeriali
...
                                         strIncludedElementType,
                                         _bValidateSchemasOnStart);
 
-        if (!fileDocSource.isParsable(elementSerializingContext)) {
+        // TODO: re-enable this
+        /* if (!docSource.isParsable(elementSerializingContext)) { */
 
-            return false;
-        }
+        /*     return false; */
```
Comment (OznOg) [inline parameter/XmlFileIncluderElement.cpp]: dead code

File: xmlserializer/XmlStreamDocSource.h [hunk 9]
```diff
@@ -64,6 +69,11 @@ class CXmlStringDocSource : public CXmlDocSource
       * @return false if any error occurs
       */
     virtual bool populate(CXmlSerializingContext& serializingContext);
+
+    static _xmlDoc* mkXmlDoc(const std::istream& input);
+
+private:
+    typedef CXmlDocSource base;
```
Comment (OznOg) [inline xmlserializer/XmlStreamDocSource.h]: Base

File: parameter/ParameterMgr.cpp (hunk 5 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 6 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: managed by an auto_ptr

File: parameter/ParameterMgr.cpp (hunk 7 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: xmlserializer/XmlStreamDocSource.h (hunk 9 above)
Comment (dawagner) [inline xmlserializer/XmlStreamDocSource.h]: done

**Comment by krocard** [comment]
:-1:

File: parameter/ParameterMgr.cpp [hunk 10]
```diff
@@ -2333,14 +2326,41 @@ bool CParameterMgr::exportDomainsXml(string& strXmlDest, bool bWithSettings, boo
...
+        /* output.reset(oss); */
+    }
+
+    // Do the actual export
+    if (!exportSingleDomainXml(*output, strDomainName, bWithSettings, strError)) {
+        return false;
+    }
+
+    if (!bToFile) {
+        strXmlDest = static_cast<ostringstream*>(output.get())->str();
```
Comment (krocard) [inline parameter/ParameterMgr.cpp]: COPY PASTE ! The whole function is nearly the same as CParameterMgr::exportDomainsXml

File: parameter/XmlFileIncluderElement.cpp (hunk 8 above)
Comment (dawagner) [inline parameter/XmlFileIncluderElement.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 10 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 2 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 3 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 4 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

**Comment by dawagner** [comment]
@freno please test this branch.

File: xmlserializer/XmlStreamDocSource.h [hunk 11]
```diff
@@ -64,6 +80,15 @@ class CXmlStringDocSource : public CXmlDocSource
...
     virtual bool populate(CXmlSerializingContext& serializingContext);
-};
 
+    /** Helper method for constructing an _xmlDoc
+     *
+     * @param[in] input an XML stream
+     *
+     * @returns an XML document object
+     */
+    static _xmlDoc* mkXmlDoc(const std::istream& input);
```
Comment (OznOg) [inline xmlserializer/XmlStreamDocSource.h]: xmlDocPtr exists (not really important)

File: parameter/ParameterMgr.cpp [hunk 12]
```diff
@@ -519,7 +536,12 @@ bool CParameterMgr::loadFrameworkConfiguration(string& strError)
     // Parse Structure XML file
     CXmlElementSerializingContext elementSerializingContext(strError);
 
-    if (!xmlParse(elementSerializingContext, getFrameworkConfiguration(), _strXmlConfigurationFilePath, _strXmlConfigurationFolderPath, EFrameworkConfigurationLibrary)) {
+    const ifstream file(_strXmlConfigurationFilePath.c_str());
+    if (!file) {
+        strError = "Failed to open " + _strXmlConfigurationFilePath + " for reading.";
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: I would add quotes around name in order to catch empty file names easily

File: parameter/ParameterMgr.cpp [hunk 13]
```diff
@@ -519,7 +536,12 @@ bool CParameterMgr::loadFrameworkConfiguration(string& strError)
...
+        return false;
+    }
+    if (!xmlParse(elementSerializingContext, getFrameworkConfiguration(), file, _strXmlConfigurationFolderPath, EFrameworkConfigurationLibrary)) {
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: fill strerror as well?

File: parameter/ParameterMgr.cpp [hunk 14]
```diff
@@ -571,7 +593,12 @@ bool CParameterMgr::loadStructure(string& strError)
 
     CAutoLog autolog(pSystemClass, "Importing system structure from file " + strXmlStructureFilePath);
 
-    if (!xmlParse(parameterBuildContext, pSystemClass, strXmlStructureFilePath, strXmlStructureFolder, EParameterCreationLibrary)) {
+    const ifstream file(strXmlStructureFilePath.c_str());
+    if (!file) {
+        strError = "Failed to open " + strXmlStructureFilePath + " for reading.";
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: idem (seems copied/pasted)

File: parameter/ParameterMgr.cpp [hunk 15]
```diff
@@ -661,7 +688,12 @@ bool CParameterMgr::loadSettingsFromConfigFile(string& strError)
     log_info("Importing configurable domains from file %s %s settings", strXmlConfigurationDomainsFilePath.c_str(), pBinarySettingsFileLocation ? "without" : "with");
 
     // Do parse
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: remove this useless comment

File: parameter/ParameterMgr.cpp [hunk 16]
```diff
@@ -1676,7 +1687,29 @@ CParameterMgr::CCommandHandler::CommandStatus
 }
 
 CParameterMgr::CCommandHandler::CommandStatus
-        CParameterMgr::getSystemClassXMLCommmandProcess(
+        CParameterMgr::setDomainWithSettingsXMLCommandProcess(
+                const IRemoteCommand& remoteCommand, string& strResult)
+{
+    bool bOverwrite = false;
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: IMHO, you should remove hungarian...

File: parameter/ParameterMgr.cpp [hunk 17]
```diff
@@ -1676,7 +1687,29 @@ CParameterMgr::CCommandHandler::CommandStatus
...
+
+    // Look for optional arguments
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: quite uselss comment

File: parameter/ParameterMgr.cpp (hunk 13 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: no need to: it is done by xmlParse through `elementSerializingContext.setError(...)`

File: parameter/ParameterMgr.cpp [hunk 18]
```diff
@@ -2190,105 +2223,85 @@ bool CParameterMgr::importDomainsXml(const string& strXmlSource, bool bWithSetti
...
-bool CParameterMgr::serializeElement(string& strXmlDest,
-                                     CXmlSerializingContext& xmlSerializingContext, bool bToFile,
-                                     const CElement& element, string& strError) const
+bool CParameterMgr::wrapLegacyXmlImport(const string& strXmlSource, bool bFromFile,
+                                        bool bWithSettings, CElement& element,
+                                        const string& strNameAttributeName, string& strError)
 {
-    // check path is absolute
-    if (bToFile && strXmlDest[0] != '/') {
+    // Context
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: useless comment

File: parameter/ParameterMgr.cpp (hunk 12 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done (in all places)

File: parameter/ParameterMgr.cpp (hunk 15 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp [hunk 19]
```diff
@@ -2351,14 +2349,37 @@ bool CParameterMgr::exportSingleDomainXml(string& strXmlDest, const string& strD
...
+
+bool CParameterMgr::wrapLegacyXmlExport(string& strXmlDest, bool bToFile, bool bWithSettings,
+                                        const CElement& element, string& strError) const
+{
+    std::auto_ptr<std::ostream> output(NULL);
+
+    if (bToFile) {
+        output.reset(new ofstream(strXmlDest.c_str()));
+        if (output->fail()) {
+            strError = "Failed to open " + strXmlDest + " for writing.";
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: quotes around name

File: parameter/ParameterMgr.cpp [hunk 20]
```diff
@@ -2351,14 +2349,37 @@ bool CParameterMgr::exportSingleDomainXml(string& strXmlDest, const string& strD
...
+        }
+    } else {
+        output.reset(new ostringstream);
+    }
+
+    CXmlDomainExportContext xmlDomainExportContext(strError, bWithSettings, _bValueSpaceIsRaw,
+                                                   _bOutputRawFormatIsHex);
 
-    xmlDomainExportContext.setValueSpaceRaw(_bValueSpaceIsRaw);
+    // Do the actual export
```
Comment (OznOg) [inline parameter/ParameterMgr.cpp]: useless comment

File: parameter/ParameterMgr.cpp (hunk 16 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.h [hunk 21]
```diff
@@ -515,35 +541,88 @@ class CParameterMgr : private CElement
...
+     * @param[in] bToFile specifies whether strXmlSource is a path or a
+     * string that will receive an actual XML document
+     * @param[in] bWithSettings if false, only export the configurations
+     * applicability rules; if true, also export their settings
+     * @param[out] element the element to be exported
+     * @param[out] strError string used as output for any error message
+     *
+     * @returns true if the export succeeded, false otherwise
+     */
+    bool wrapLegacyXmlExport(std::string& strXmlDest, bool bToFile, bool bWithSettings,
```
Comment (OznOg) [inline parameter/ParameterMgr.h]: why hungarian...

File: xmlserializer/XmlDocSource.cpp [hunk 22]
```diff
@@ -133,6 +129,13 @@ _xmlDoc* CXmlDocSource::getDoc() const
     return _pDoc;
 }
 
+bool CXmlDocSource::isParsable() const
+{
+    // Check that the doc has been created
+    // Implicit pointer->bool conversion
+    return _pDoc;
```
Comment (OznOg) [inline xmlserializer/XmlDocSource.cpp]: _pDoc != NULL

File: xmlserializer/XmlDocSource.cpp [hunk 23]
```diff
@@ -133,6 +129,13 @@ _xmlDoc* CXmlDocSource::getDoc() const
     return _pDoc;
 }
 
+bool CXmlDocSource::isParsable() const
+{
+    // Check that the doc has been created
+    // Implicit pointer->bool conversion
```
Comment (OznOg) [inline xmlserializer/XmlDocSource.cpp]: well please remove this useless comment, I'm not sure this is the right place to learn C++

File: xmlserializer/XmlStreamDocSink.cpp [hunk 24]
```diff
@@ -54,7 +52,7 @@ bool CXmlStringDocSink::doProcess(CXmlDocSource& xmlDocSource,
         return false;
     }
 
-    _strResult.append((const char*)pcDumpedDoc);
+    _output << (const char*)pcDumpedDoc;
```
Comment (OznOg) [inline xmlserializer/XmlStreamDocSink.cpp]: nice exmaple of the EVIL of hungarian: this is NOT a pcXXX this is a xmlChar... and despite the guess work, it is not a char *....

File: parameter/ParameterMgr.cpp (hunk 17 above)
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: done

File: xmlserializer/XmlStreamDocSink.cpp [hunk 25]
```diff
@@ -27,36 +27,34 @@
...
-    if (!pcDumpedDoc) {
+    if (!dumpedDoc) {
 
         serializingContext.setError("Unable to encode XML document in memory");
 
         return false;
     }
 
-    _strResult.append((const char*)pcDumpedDoc);
+    _output << (const char*)dumpedDoc;
```
Comment (krocard) [inline xmlserializer/XmlStreamDocSink.cpp]: Are you sure you can safely cast `xmlChar*` to `const char *`. Maybe use a static_cast? I locked at the doc, it seems xmlChar is a typedef of unsined char, as a result no cast seems needed.

File: xmlserializer/XmlDocSource.cpp [hunk 26]
```diff
@@ -133,6 +101,18 @@ _xmlDoc* CXmlDocSource::getDoc() const
     return _pDoc;
 }
 
+bool CXmlDocSource::isParsable() const
```
Comment (krocard) [inline xmlserializer/XmlDocSource.cpp]: `hasBeenParsed` or `wasParsed`?

Comment (dawagner) [inline xmlserializer/XmlDocSource.cpp]: was already named like this.

**Comment by dawagner** [comment]
fixed a typo in a commit message

**Pull request merged**